File: LeetCode/350两个数组的交集Ⅱ.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            return intersect(nums2, nums1);
        }
        Map<Integer, Integer> map = new HashMap<>();
        List<Integer> list = new ArrayList<>();
        for (int i = 0; i < nums1.length; i++) {
            map.put(nums1[i], map.getOrDefault(nums1[i], 0) + 1);
        } 
        for (int i = 0; i < nums2.length; i++) {
            int count = map.getOrDefault(nums2[i], 0);
            if (count > 0) {
                list.add(nums2[i]);
                count--;
                if (count > 0) {
                    map.put(nums2[i], count);
                } else {
                    map.remove(nums2[i]);
                }
            }
        }
        int[] ans = list.stream().mapToInt(i->i).toArray();
        return ans;
    }
}

File: LeetCode/350两个数组的交集Ⅱ.go
func intersect(nums1 []int, nums2 []int) []int {
	if len(nums1) > len(nums2) {
		intersect(nums2, nums1)
	}
	hash := map[int]int{}
	var ans []int
	for _, v := range nums1 {
		hash[v]++
	}
	for _, v := range nums2 {
		cnt := hash[v]
		if cnt > 0 {
			ans = append(ans, v)
			cnt--
			if cnt > 0 {
				hash[v] = cnt
			} else {
				delete(hash, v)
			}
		}
	}
	return ans
}